include/triton-shared/Conversion/TritonArithToLinalg/ConversionPatterns.hpp

```diff
@@ -150,6 +150,36 @@ static std::optional<unsigned> getBitWidth(Type a) {
   return std::nullopt;
 }
 
+static void processTwoStridesLastDim(RankedTensorType type, ConversionPatternRewriter &rewriter,
+    std::function<void(SmallVector<OpFoldResult, 4> &, SmallVector<OpFoldResult, 4>&, SmallVector<OpFoldResult, 4>&, int)> op) {
+  int64_t rank = type.getRank();
+  auto shape = type.getShape();
+
+  SmallVector<OpFoldResult, 4> offsets, sizes, strides;
+  std::vector<int64_t> stridesInt(rank, 0);
+
+  int64_t currentStride = 1;
+  for (size_t i = shape.size(); i > 0; --i) {
+    stridesInt[i - 1] = currentStride;
+    currentStride *= stridesInt[i - 1];
+  }
+
+  for (size_t j = 0; j < shape.size(); ++j) {
+    offsets.push_back(rewriter.getIndexAttr(0));
+    sizes.push_back(rewriter.getIndexAttr(shape[j]));
+    strides.push_back(rewriter.getIndexAttr(stridesInt[j]));
+  }
+
+  for (int i = 0; i < 2; ++i) {
+    offsets.pop_back();
+    sizes.pop_back();
+
+    offsets.push_back(rewriter.getIndexAttr(i));
+    sizes.push_back(rewriter.getIndexAttr(1));
+    op(offsets, sizes, strides, i);
+  }
+}
+
```

Comment:
I have a few nit / cosmetic comments about this function:
+ [ ] Should probably rename `processTwoStridesLastDim` to something else. The implementation looks rather fiddly too: are we replacing values of the last 2 dimensions with the first 2 dimensions?
+ [ ] The lambda's arguments should probably have names so that we can better understand what each argument is for.
+ [ ] nit: Our team prefers not to put the size in `SmallVector` declaration
+ [ ] Mixed usage of vector and SmallVector

Reply:
1. sure, the function assumes there are only two strides for the last dimension and it applies given lambda on these two strides, I have the following options for the name: `operateOnSplitStrides`, `applyToTensorStrides`. Let me know what you think.

Pop-push is there to reuse all data structures between two strided calls, I could evaluate them independently, but that would mean redundant computation for all the dimensions except the last one. What it does now, it computes everything for the whole tensor, then discards the last offset and size, replaces it with 0, 1 for the first stride and 1,1 for the second.
2. done, put names as comments
3. sure, done
4. sorry, fixed

Reply:
Thank you for resolving the issues. As for `processTwoStridesLastDim`, I still have some questions.

> sure, the function assumes there are only two strides for the last dimension and it applies given lambda on these two strides

The part where the last dimension contains two strides is still not clear to me. Why are there two strides for a single dimension?

> What it does now, it computes everything for the whole tensor, then discards the last offset and size, replaces it with 0, 1 for the first stride and 1,1 for the second.

This is also quite confusing to me. What is the `index` in the lambda supposed to mean? In each loop iteration we're only changing the *last* value of offsets and sizes, and yet the index is `0` and `1`.

I understand the function is trying to reuse code in both `Split` and `Join`, but it's still quite hard for me to grasp exactly what it's trying to encapsulate here.

Reply:
Yes, index here is offset for example, we have a tensor of <128x2xf16>, this lambda runs given function two times with:
(size = (128, 1), offset = (0, 0), stride = (1, 1))
(size = (128, 1), offset = (0, 1), stride = (1, 1))
we always iterate over the big tensor which makes it two strides in this case.

This function I use in one case to copy two small tensors into the big one with extra dimension at the end (which I iterate over). In the second case I copy from this big tensor into the small ones. Since the code for these two operations are mostly the same I put it as a function. Do not hesitate to ask any follow-up questions, I hope that my answers are helping.